### ISSUE TITLE
bug: spreading `Headers` object has breaking behavior in versions between Node 20.12 and 20.13

### ISSUE THREAD

**Issue by juliusmarminge** [nodejs/undici#3267]
## Reproducible By

```js
// Comment out this comment and have the latest undici installed just to verify it's still
// present in latest Undici, but for the test you can use the global Undici bundled in Node
// import { fetch } from "undici";

{
  const baseHeaders = { "x-foo": "bar" };

  const requestHeaders = new Headers({ "Content-Type": "application/json" });
  const headers = {
    ...baseHeaders,
    ...requestHeaders,
  };
  console.log("Result", headers);

  await fetch("https://google.com", {
    method: "POST",
    headers,
  });
}
```

Run the code in Node 20.12 and 20.13. 20.12 works fine, while 20.13 panics with `TypeError: Could not convert argument of type symbol to string`.

## Expected Behavior

It to work like it did in 20.12 - afterall I didn't bump any major version, so if this is an intentional breaking change on Undici's side it shouldn't have been included in a minor version of Node.js.

## Logs & Screenshots

![CleanShot 2024-05-16 at 12 00 10](https://github.com/nodejs/undici/assets/51714798/7112cd53-4a75-4a81-a98e-d3e363edee53)

## Environment

- Latest MacOS
- Node 20.13

### Additional context

A separate issue for Node perhaps, but spreading the Headers object probably shouldn't even include the Symbol(headers list) to begin with, this is how it looks in the browser for example:

![CleanShot 2024-05-16 at 12 04 29](https://github.com/nodejs/undici/assets/51714798/9e90bfd8-f771-4578-bd97-24b9144e50dd)

**Comment by kettanaito** [comment]
Reported tangentially by MSW users as well.

**Comment by tsctx** [comment]
You are using the spread operator incorrectly.
In `Headers` you need to use it with arrays, not objects.

**Comment by tsctx** [comment]
just use `Object.fromEntries`.

```js
// Comment out this comment and have the latest undici installed just to verify it's still
// present in latest Undici, but for the test you can use the global Undici bundled in Node
// import { fetch } from "undici";

{
  const baseHeaders = { "x-foo": "bar" };

  const requestHeaders = new Headers({ "Content-Type": "application/json" });
  const headers = {
    ...baseHeaders,
    ...Object.fromEntries(requestHeaders),
  };
  console.log("Result", headers);

  await fetch("https://google.com", {
    method: "POST",
    headers,
  });
}
```

**Comment by juliusmarminge** [comment]
Yes I know I can do that, (and I did work around the issue in https://github.com/pingdotgg/uploadthing/pull/813) - but that's working around the issue that the behavior changed in a minor Node.js release

**Comment by mcollina** [comment]
From a quick look at the behaviors in browsers, I think that the correct implementation is:

```js
  const baseHeaders = { "x-foo": "bar" };

  const requestHeaders = new Headers({ "Content-Type": "application/json" });
  const headers = {
    ...baseHeaders,
    ...Object.fromEntries(requestHeaders),
  };
```

Note that the following does not work in browsers either:

```js
  const baseHeaders = { "x-foo": "bar" };

  const requestHeaders = new Headers({ "Content-Type": "application/json" });
  const headers = {
    ...baseHeaders,
    ...requestHeaders,
  };
  console.log("Result", headers);
```

It will result in not having `Content-Type` in `headers` object.

Having said that, I think it should not crash but works as expected & similarly to browsers.

---

@KhafraDev ping

**Comment by kettanaito** [comment]
@mcollina, yeah, you are right. Headers instance itself won't spread, it will _return an empty object_. But neither will it throw, so that behavior is unexpected although useful as it lets the user know they aren't doing anything (the error is not hinting at that, though).

**Comment by juliusmarminge** [comment]
> It will result in not having Content-Type in headers object

Oh wow ai didnt even realize that. Was too focused looking at the Symbols xD

**Comment by juliusmarminge** [comment]
Checked some other runtimes:

![CleanShot 2024-05-16 at 13 02 02](https://github.com/nodejs/undici/assets/51714798/d54d58f7-
2cfb-406b-9f8a-37525a0af56b)

![CleanShot 2024-05-16 at 13 04 17](https://github.com/nodejs/undici/assets/51714798/2bcff843-8467-4725-8619-4917e4278564)

**Comment by KhafraDev** [comment]
This was caused by fd66cf04b222ee365cdca780d9b465f531f74eb5 which exposed the issue. Our only choice is to move towards private properties I think.

**Comment by File-permissionsagreed26** [comment]
https://angular-ui-bootstrap-tab.github.io/2rEtfK3Ip2gpOHe1pzqsGWlSFPLMj2/
